ocaml/xapi/repository_helpers.ml

```diff
@@ -174,56 +174,48 @@ end
-    match Uri.scheme uri with
-    | Some "http" | Some "https" -> (
-      match Uri.host uri with
-      | Some h -> (
-          let subdomain =
-            DomainNameIncludeIP.(host_exn (of_string h |> Result.get_ok))
-          in
-          let valids, invalids =
-            List.fold_left
-              (fun (valids, invalids) d ->
-                match Domain_name.of_string d with
-                | Ok domain ->
-                    (domain :: valids, invalids)
-                | Error _ ->
-                    (valids, d :: invalids)
+    match Uri.(scheme uri, host uri) with
+    | Some ("http" | "https"), Some h -> (
+        let subdomain =
+          DomainNameIncludeIP.(host_exn (of_string h |> Result.get_ok))
+        in
+        let invalids, valids =
+          List.partition_map
+            (fun d ->
+              match Domain_name.of_string d with
+              | Ok dom ->
+                  Right dom
+              | Error _ ->
+                  Left d
+            )
+            !Xapi_globs.repository_domain_name_allowlist
+        in
+        let hostname_allowed =
+          List.exists (fun d ->
+              DomainNameIncludeIP.(is_subdomain ~subdomain ~domain:(host_exn d))
+          )
+        in
+        match (valids, invalids) with
+        | [], [] ->
+            ()
+        | valids, [] when not (hostname_allowed valids) ->
+            let msg = "host is not in allowlist" in
+            raise Api_errors.(Server_error (internal_error, [msg]))
+        | _, [] ->
+            ()
+        | _ ->
+            error "Invalid domains: %s in repository-domain-name-allowlist."
+              (String.concat "," invalids) ;
+            raise
+              Api_errors.(
+                Server_error (invalid_repository_domain_allowlist, invalids)
               )
-              ([], [])
-              !Xapi_globs.repository_domain_name_allowlist
-          in
-          match (valids, invalids) with
-          | [], [] ->
-              ()
-          | _, invalids when invalids <> [] ->
-              error "Invalid domains: %s in repository-domain-name-allowlist."
-                (String.concat "," invalids) ;
-              raise
-                Api_errors.(
-                  Server_error (invalid_repository_domain_allowlist, invalids)
-                )
-          | valids, _ when valids <> [] ->
-              let hostname_allowed =
-                List.exists (fun d ->
-                    DomainNameIncludeIP.(
-                      is_subdomain ~subdomain ~domain:(host_exn d)
-                    )
-                )
-              in
-              if not (hostname_allowed valids) then
-                let msg = "host is not in allowlist" in
-                raise Api_errors.(Server_error (internal_error, [msg]))
-          | _ ->
-              ()
-        )
-      | None ->
-          raise
-            Api_errors.(Server_error (internal_error, ["invalid host in url"]))
-    )
+      )
     | _ ->
```

Comment:
The distinction between invalid host and invalid scheme should be kept. This can be done by changing the case to two different ones using
```ocaml 
    | _, None ->
          host is incorrect
    | (Some _ | None), _ ->
          scheme is incorrect
```